test/Microsoft.AspNet.Mvc.TagHelpers.Test/ScriptTagHelperTest.cs

```diff
@@ -0,0 +1,204 @@
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(1)]
+        public async Task DoesNotRunWhenARequiredAttributeIsMissing(int attributeToRemove)
+        {
+            // Arrange
+            var attributes = new Dictionary<string, object>
+                {
+                    { "asp-fallback-src", "http://www.example.com/blank.js" },
+                    { "asp-fallback-test", "isavailable()" },
+                };
+
+            var removed = attributes.Keys.ElementAt(attributeToRemove);
+            attributes.Remove(removed);
+
+            Assert.Equal(1, attributes.Count);
+
+            var context = MakeTagHelperContext(attributes);
+
+            var output = MakeTagHelperOutput("script");
+            var logger = new Logger();
+
+            // Act
+            var helper = new ScriptTagHelper
+            {
+                Logger = logger,
+            };
+
+            await helper.ProcessAsync(context, output);
+
+            // Assert
+            Assert.NotNull(output.TagName);
+            Assert.False(output.ContentSet);
+
+            Assert.Equal(2, logger.Logged.Count);
+
+            Assert.Equal(LogLevel.Warning, logger.Logged[0].Item1);
+            Assert.IsType<MissingAttributeLoggerStructure>(logger.Logged[0].Item2);
+
+            var loggerData0 = (MissingAttributeLoggerStructure)logger.Logged[0].Item2;
+            Assert.Equal(1, loggerData0.MissingAttributes.Count());
+            Assert.Equal(removed, loggerData0.MissingAttributes.Single());
+
+            Assert.Equal(LogLevel.Verbose, logger.Logged[1].Item1);
+            Assert.IsType<string>(logger.Logged[1].Item2);
+            Assert.StartsWith("Skipping processing for ScriptTagHelper", (string)logger.Logged[1].Item2);
+        }
+
+        [Fact]
+        public async Task DoesNotRunWhenAllRequiredAttributesAreMissing()
+        {
+            // Arrange
+            var context = MakeTagHelperContext();
+            var output = MakeTagHelperOutput("script");
+            var logger = new Logger();
+
+            // Act
+            var helper = new ScriptTagHelper
+            {
+                Logger = logger,
+            };
+
+            await helper.ProcessAsync(context, output);
+
+            // Assert
+            Assert.NotNull(output.TagName);
+            Assert.False(output.ContentSet);
+
+            Assert.Equal(1, logger.Logged.Count);
+
+            Assert.Equal(LogLevel.Verbose, logger.Logged[0].Item1);
+            Assert.IsType<string>(logger.Logged[0].Item2);
+            Assert.StartsWith("Skipping processing for ScriptTagHelper", (string)logger.Logged[0].Item2);
+        }
+
+        [Fact]
+        public async Task PreservesOrderOfSourceAttributesWhenRun()
+        {
+            // Arrange
+            var context = MakeTagHelperContext(
+                attributes: new Dictionary<string, object>
+                {
+                    { "data-extra", "something"},
```

Comment:
This would be better if the two attributes you expect to be in the output were not already in alphabetical order. Perhaps add a third and put them in non-alphabetical order.